base/sparse/sparsematrix.jl

```diff
@@ -1433,7 +1433,7 @@ function _broadcast_unary_nz2z_z2z_T{TvA,TiA,TvB}(f::Function, A::SparseMatrixCS
     return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
 end
 function _broadcast_unary_nz2z_z2z{Tv}(f::Function, A::SparseMatrixCSC{Tv})
-    _broadcast_unary_nz2z_z2z_T(f, A, Tv)
+    _broadcast_unary_nz2z_z2z_T(f, A, promote_op(f, Tv))
```

Comment:
Even for `nz2z_z2z` you could use `broadcast(f, A.nzval)` and filter out zeros.  You could even do it in-place (with a `resize!` at the end).